### PR TITLE
Integration/visio plainte

### PR DESCRIPTION
**Pull request by lebaudantoine**

### CODE REVIEW

File: src/backend/core/tests/services/test_transit_code.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+"""
```
Comment (lebaudantoine) [inline src/backend/core/tests/services/test_transit_code.py:1]: Add a test if the cache delete fails to make sure it returns None.

File: src/backend/core/tests/test_external_api_users.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    token = generate_test_token(user, [ApplicationScope.USERS_SESSION])
+
+    client = APIClient()
+    client.credentials(HTTP_AUTHORIZATION=f"Bearer {token}")
+    response = client.post("/external-api/v1.0/users/transit-code/")
+
+    assert response.status_code == 401
+
+
+# todo - add a test to make sure the addon authentification doesn't allow to mint a transit token
```
Comment (lebaudantoine) [inline src/backend/core/tests/test_external_api_users.py]: Add a test to make sure GET are forbidden.

File: src/backend/core/external_api/viewsets.py [hunk 3]
```diff
@@ -218,3 +219,62 @@ def perform_create(self, serializer):
                 "$set": {"email": self.request.user.email},
             },
         )
+
+
+class UserViewSet(viewsets.GenericViewSet):
```
Comment (lebaudantoine) [inline src/backend/core/external_api/viewsets.py:228]: sensitive viewset, discuss whether adding throttling, per client_id and per user_id

File: src/backend/core/tests/test_api_users_exchange_access_token.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    # `secrets` rather than `random`: the global random module is seeded
+    # deterministically by the factories, its sequence repeats across runs.
+    remote_addr = (
+        f"10.{secrets.randbelow(256)}.{secrets.randbelow(256)}"
+        f".{secrets.randbelow(254) + 1}"
+    )
+    return APIClient(REMOTE_ADDR=remote_addr)
+
+
+def test_exchange_access_token_missing_code(client):
```
Comment (lebaudantoine) [inline src/backend/core/tests/test_api_users_exchange_access_token.py:52]: make sure GET are forbidden

File: src/backend/core/api/serializers.py [hunk 5]
```diff
@@ -599,3 +604,25 @@ class ExternalProcessEventSerializer(BaseValidationOnlySerializer):
...
+
+    def validate_code(self, value):
+        """Reject codes whose length cannot match a generated one.
+
+        `secrets.token_urlsafe(nbytes)` produces (4 * nbytes + 2) // 3
+        url-safe characters. Checking the length against the configured
+        TRANSIT_CODE_NBYTES makes malformed codes fail fast with a 400,
+        before any cache lookup.
+        """
+        expected_length = (4 * settings.TRANSIT_CODE_NBYTES + 2) // 3
```
Comment (lebaudantoine) [inline src/backend/core/api/serializers.py:618]: verify the math

File: src/backend/core/tests/rooms/test_api_rooms_subtitle.py [hunk 6]
```diff
@@ -219,10 +219,13 @@ def test_start_subtitle_wrong_signature(settings, mock_livekit_token):
...
+        HTTP_AUTHORIZATION=f"X-LiveKit-Token {mock_livekit_token}",
     )
 
     assert response.status_code == 403
     assert response.json() == {
         "detail": "Invalid LiveKit token: Signature verification failed"
     }
+
+
+# todo - try to pass another scheme to make sure it defers to the next auth
```
Comment (lebaudantoine) [inline src/backend/core/tests/rooms/test_api_rooms_subtitle.py]: try to pass a bearer as a regression test and assert the result

File: src/backend/core/tests/rooms/test_api_rooms_rename_toggle.py [hunk 7]
```diff
@@ -565,10 +611,16 @@ def test_rename_participant_not_found(mock_livekit_client, room, token):
...
+        HTTP_AUTHORIZATION=f"X-LiveKit-Token {token}",
     )
 
     assert response.status_code == status.HTTP_404_NOT_FOUND
     assert response.data == {"error": "Participant not found"}
 
     mock_livekit_client.aclose.assert_called_once()
+
+
+# todo - try to pass another scheme to make sure it defers to the next auth
```
Comment (lebaudantoine) [inline src/backend/core/tests/rooms/test_api_rooms_rename_toggle.py]: try to pass a bearer as a regression test and assert the result

File: src/backend/core/tests/rooms/test_api_rooms_participants_management.py [hunk 8]
```diff
@@ -1020,3 +1032,6 @@ def test_remove_participant_not_found(mock_livekit_client):
     assert response.data == {"error": "Participant not found"}
 
     mock_livekit_client.aclose.assert_called_once()
+
+
+# todo - try to pass another scheme to make sure it defers to the next auth
```
Comment (lebaudantoine) [inline src/backend/core/tests/rooms/test_api_rooms_participants_management.py]: try to pass a bearer as a regression test and assert the result

File: src/backend/core/tests/rooms/test_api_rooms_retrieve.py [hunk 9]
```diff
@@ -507,3 +510,40 @@ def test_api_rooms_retrieve_administrators(
...
+    room = RoomFactory(users=[(user, "owner")])
+
+    client = APIClient()
+    client.credentials(HTTP_AUTHORIZATION=f"Bearer {generate_user_access_token(user)}")
+    response = client.get(f"/api/v1.0/rooms/{room.id!s}/")
+
+    assert response.status_code == 200
+    assert response.data["id"] == str(room.id)
+    # Authenticated as the owner: privileged fields are included
+    assert response.data["pin_code"] == room.pin_code
```
Comment (lebaudantoine) [inline src/backend/core/tests/rooms/test_api_rooms_retrieve.py:554]: todo check if more "owner" fields should be included 
Also add a test with a user not owner, probably with no role on the room to make sure the access token don't bypass the permission control

**Review by lebaudantoine — COMMENTED** [review]
wip

File: src/backend/core/api/throttling.py [hunk 10]
```diff
@@ -58,10 +59,14 @@ def get_cache_key(self, request, view):
...
+        serializer = serializers.RequestEntrySerializer(data=request.data)
+        if not serializer.is_valid():
+            return None
 
-        if participant_id is None:
-            return None  # No throttling for cookieless requests
+        participant_id = serializer.validated_data.get("participant_id")
+
+        if not participant_id:
+            return None  # No throttling for unidentified requests
```
[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/backend/core/api/throttling.py:69]: not appropriated, users are behind NAT

File: src/backend/core/services/lobby.py [hunk 11]
```diff
@@ -194,52 +178,48 @@ def request_entry(
...
-                    id=participant_id,
-                    color=utils.generate_color(participant_id),
-                    entered_at=timezone.now().isoformat(),
-                )
-            else:
-                participant.status = LobbyParticipantStatus.ACCEPTED
+        user_role = room.get_role(user)
+
+        if self.can_bypass_lobby(room=room, user=user, role=user_role):
+            participant = self.handle_participant_entry(room_id, participant.id, True)
```
Comment (lebaudantoine) [inline src/backend/core/services/lobby.py]: Changes I introduced: code now saves lobby-bypassing participants as ACCEPTED. Subsequent requests trust that cached status even when the caller no longer qualifies for admission. An anonymous visitor can obtain an ID while the room is public, then reuse it after the room becomes restricted to obtain fresh LiveKit tokens without approval, until the cached acceptance expires (default: six hours). The base version puts that caller into the waiting lobby.

File: src/frontend/src/stores/lobby.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+) => {
+  layoutStore.participantIds[roomId] = participantId
+}
+
+export const clearParticipantId = (roomId: string) => {
+  delete layoutStore.participantIds[roomId]
+}
+
+export const getLobbyParticipantId = (roomId: string) =>
+  layoutStore.participantIds[roomId]
```
Comment (lebaudantoine) [inline src/frontend/src/stores/lobby.ts:23]: should be synced with session storage

File: src/backend/core/services/lobby.py [hunk 13]
```diff
@@ -131,23 +131,6 @@ def _index_remove(self, room_id: UUID, *participant_ids: str) -> None:
...
-        """Set participant cookie if needed."""
-        if not response.cookies.get(settings.LOBBY_COOKIE_NAME):
-            response.set_cookie(
-                key=settings.LOBBY_COOKIE_NAME,
-                value=participant_id,
-                httponly=True,
-                secure=True,
-                samesite="Lax",
-            )
-
```
Comment (lebaudantoine) [inline src/backend/core/services/lobby.py:150]: let's discuss having an http-only cookie for first party usage, and a payload one for embedded scenario

File: src/backend/meet/settings.py [hunk 14]
```diff
@@ -1025,6 +1031,66 @@ class Base(Configuration):
...
+        environ_prefix=None,
+    )
+    # Number of random bytes per code (48 bytes -> 64 url-safe characters)
+    TRANSIT_CODE_NBYTES = values.PositiveIntegerValue(
+        48,
+        environ_name="TRANSIT_CODE_NBYTES",
+        environ_prefix=None,
+    )
+    USER_ACCESS_TOKEN_TYPE = values.Value(
+        "Bearer",
```
Comment (briquet) [inline src/backend/meet/settings.py:1085]: There are a few places in `BaseJWTAuthentication` where the the token type is hardcoded 
- [authenticate_header](https://github.com/suitenumerique/meet/blob/main/src/backend/core/external_api/authentication.py#L162)
- [authenticate](https://github.com/suitenumerique/meet/blob/main/src/backend/core/external_api/authentication.py#L76)

It might break if the settings is changed from it's default value `Bearer`

File: src/backend/core/services/transit_code.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    def _cache_key(code):
+        """Build the cache key for a code.
+
+        The code is hashed so that a dump of the cache never reveals
+        directly usable codes.
+        """
+        digest = hashlib.sha256(code.encode("utf-8")).hexdigest()
+        return f"{settings.TRANSIT_CODE_CACHE_PREFIX}:{digest}"
+
+    def create_code(self, user, client_id="unknown"):
```
Comment (briquet) [inline src/backend/core/services/transit_code.py:30]: The `client_id` could be mandatory and validated with a serializer the `UserViewSet`

File: src/backend/core/tests/rooms/test_api_rooms_create.py [hunk 16]
```diff
@@ -312,3 +315,39 @@ def test_api_rooms_create_authenticated_blank_user_default_access_level():
     assert response.status_code == 201
     room = Room.objects.get()
     assert room.access_level == settings.RESOURCE_DEFAULT_ACCESS_LEVEL
+
+
+def generate_user_access_token(user):
```
Comment (briquet) [inline src/backend/core/tests/rooms/test_api_rooms_create.py:320]: You could extract the `generate_user_access_token` in `src/backend/core/tests/test_api_user_access_token_authentication.py` as an helper because its very similar and also used in [test_api_rooms_list.py](src/backend/core/tests/rooms/test_api_rooms_list.py) + [test_api_rooms_retrieve.py](src/backend/core/tests/rooms/test_api_rooms_retrieve.py)

File: src/backend/core/api/viewsets.py [hunk 17]
```diff
@@ -219,6 +222,96 @@ def get_me(self, request):
...
+        if code_data is None:
+            logger.warning("Invalid, expired or already used transit code")
+            raise drf_exceptions.PermissionDenied(
+                "Invalid, expired or already used transit code."
+            )
+
+        # Re-check the user at exchange time so that a deactivation after
+        # the transit code was minted is taken into account.
+        try:
+            user = models.User.objects.get(id=code_data["user_id"], is_active=True)
```
Comment (briquet) [inline src/backend/core/api/viewsets.py:269]: nit: you could extract the `client_id, user_id` from `code_data` because it's used many times

File: src/backend/core/authentication/user_token.py [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    def __init__(self):
+        """Initialize the backend with user access token settings."""
+        super().__init__(
+            secret_key=settings.USER_ACCESS_TOKEN_SECRET_KEY,
+            algorithm=settings.USER_ACCESS_TOKEN_ALG,
+            issuer=settings.USER_ACCESS_TOKEN_ISSUER,
+            audience=settings.USER_ACCESS_TOKEN_AUDIENCE,
+            expiration_seconds=settings.USER_ACCESS_TOKEN_TTL,
+            token_type=settings.USER_ACCESS_TOKEN_TYPE,
+            is_enabled=settings.USER_ACCESS_TOKEN_ENABLED,
```
Comment (briquet) [inline src/backend/core/authentication/user_token.py:53]: The feature flag activation is a bit hidden inside the parent class:  you could modify the `DEFAULT_AUTHENTICATION_CLASSES` settings based on the value of `settings.USER_ACCESS_TOKEN_ENABLED` using django-configuration `@property`

File: src/backend/core/api/viewsets.py [hunk 19]
```diff
@@ -219,6 +222,96 @@ def get_me(self, request):
...
+
+        code_data = TransitCodeService().consume_code(serializer.validated_data["code"])
+
+        if code_data is None:
+            logger.warning("Invalid, expired or already used transit code")
+            raise drf_exceptions.PermissionDenied(
+                "Invalid, expired or already used transit code."
+            )
+
+        # Re-check the user at exchange time so that a deactivation after
```
Comment (briquet) [inline src/backend/core/api/viewsets.py:266]: Is it really necessary ? 

The minted JWT token validation always check whether the user is active in `UserAccessJWTAuthentication.get_user` when validating the JWT token.

File: src/backend/core/tests/rooms/test_api_rooms_update.py [hunk 20]
```diff
@@ -410,3 +414,46 @@ def test_api_rooms_update_livekit_sync_failure(mock_update_metadata, exception):
...
+    room = RoomFactory(users=[(user, "member")])
+
+    client = APIClient()
+    client.credentials(HTTP_AUTHORIZATION=f"Bearer {generate_user_access_token(user)}")
+
+    # A simple member cannot update the room
+    response = client.patch(f"/api/v1.0/rooms/{room.id!s}/", {"name": "new name"})
+    assert response.status_code == 403
+
+    # An administrator can
```
Comment (briquet) [inline src/backend/core/tests/rooms/test_api_rooms_update.py:454]: Shouldn't we test the owner as well ?

File: src/backend/core/authentication/user_token.py [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+from django.conf import settings
+
+from rest_framework import exceptions
+
+from core.external_api.authentication import BaseJWTAuthentication
+from core.models import Application, ApplicationScope
+
+logger = logging.getLogger(__name__)
+
+USER_ACCESS_TOKEN_TYPE_CLAIM = "user_access"  # noqa: S105
```
Comment (briquet) [inline src/backend/core/authentication/user_token.py:26]: Why not using the `settings.USER_ACCESS_TOKEN_TYPE_CLAIM` ?

**Comment by briquet** [comment]
LGTM :+1: 
I would have use classic `postMessage` instead of URL fragment but the implementation is simpler to maintain
